tf_quant_finance/models/heston/approximations/asian_prices.py

```diff
@@ -0,0 +1,509 @@
+    x = tf.math.sqrt(tf.math.pow(self.mean_reversion, 2)
+                     - 2 * z1 * tf.math.pow(self.volvol, 2))
+
+    term_1 = 0.5 * x * tf.math.sinh(0.5 * tau * x)
+    term_2 = 0.5 * ((self.mean_reversion - z2 * tf.math.pow(self.volvol, 2))
+                     * tf.math.cosh(0.5 * tau * x))
+
+    return term_1 + term_2
+
+  def a(self, s, w):
+    """Calculates a(s, w) (eq. (17) from the referenced paper)"""
+    raw_drift = self.discount_rates - self.dividend_rates
+    corr_drift = self.rho * self.mean_reversion * self.theta / self.volvol
+    mod_drift = raw_drift - corr_drift
+
+    summed_samplings = tf.expand_dims(
+      tf.reduce_sum(self.sampling_times[self.k_star:], axis=0),
+      -1
+    )
+
+    n_c = tf.constant(self.n, dtype=self.dtype) 
+    k_star_c = tf.constant(self.k_star, dtype=self.dtype)
+
+    term_1 = ((s * (n_c - k_star_c) / n_c + w)
+               * (tf.math.log(self.spots)
+                  - self.rho * self.variances / self.volvol
+                  - self.start_times * mod_drift))
+    term_2 = (s * (summed_samplings) / n_c + w * self.expiries) * mod_drift
+
+    return term_1 + term_2
+
+  def z(self, s, w):
+    """Calculates z_k(s, w) (eq. (14) from the referenced paper)"""
+    k_s_w_tensor = tf.expand_dims(self.n - self.k_tensor + 1, 2) * s + self.n * w
+    term_1 = ((2 * self.rho * self.mean_reversion - self.volvol)
+               * (k_s_w_tensor) / (self.volvol * self.n))
+    term_2 = ((1 - tf.math.pow(self.rho, 2))
+               * tf.cast(tf.math.pow(k_s_w_tensor, 2), dtype=self.dtype)
+               / tf.cast(tf.math.pow(self.n, 2), dtype=self.dtype))
+
+    return 0.5 * (term_1 + term_2)
+
+  def omega_k(self, s, w):
+    """Calculates \omega_k(s, w) (eq. (15) from the referenced paper)"""
+    return tf.concat([tf.expand_dims(self.start_times, 1) * s,
+                      tf.expand_dims(tf.ones((self.sampling_times.shape[1], self.n - self.k_star), dtype=self.dtype) * self.rho / (self.volvol * self.n), 2) * s,
+                      tf.expand_dims(self.rho / self.volvol, 1) * w], 1)
+
+  def omega_tilde_k(self, s, w, z_k_tensor):
+    """Calculates \tilde{\omega}_k(s, w) (eq. (19) from the referenced paper)"""
+    omega_tilde = self.omega_k(s, w)
+    mr_v2 = self.mean_reversion / tf.math.pow(self.volvol, 2)
+
+    rows = []
+    last_row = None
+
+    for q in reversed(range(0, omega_tilde.shape[1])):
+        row = omega_tilde[:,q,:]
+
+        if last_row is not None:
+            d_tau = tf.expand_dims(self.d_tau_k[q], 1)
+            z_k = z_k_tensor[q]
+
+            row += mr_v2 - (2 * self.F_tilde(z_k, last_row, d_tau)
+                            / (self.F(z_k, last_row, d_tau)
+                            * tf.math.pow(self.volvol, 2)))
+
+        rows.append(row)
+        last_row = row
+
+    rows = list(reversed(rows))
+    omega_tilde = tf.stack(rows)
+
+    return omega_tilde
+
+  def Phi(self, s, w):
```

Comment:
nit: we'd need to follow style guide  and call it `phi`

Reply:
I thought a bit about this and went with `Phi` instead of `phi` in the end because it refers to a function in the reference paper that the authors call \Phi.

Although lowercase functions are preferred in the PEP8 style guide, it also says that for the sake of clarity and explicitness we can occasionally break the rules.

I'm happy to defer if you have very strong feelings on this but I prefer matching the reference here.

Reply:
internally we have to follow the style guide. Could you please rename to `phi` and comment that this is the same as `Phi`?

Reply:
Renamed